### ISSUE TITLE
Incorrect generic return type not checked in staticmethod

### ISSUE THREAD

**Issue by kwshi** [microsoft/pyright#2709]
**Describe the bug**
In certain situations involving a generic class's `@staticmethod` returning an instance of the same class, the returned class's generic parameter type is not correctly checked.

**To Reproduce**
Consider the following generic class with three equivalent functions returning instances of itself, with incorrect parameter types.
```python
import typing
from __future__ import annotations

T = typing.TypeVar("T")


class Banana(typing.Generic[T]):
    stuff: T

    def __init__(self, thing: T):
        self.stuff = thing

    # NO ERROR (unexpected behavior)
    @staticmethod
    def x(_: T) -> Banana[T]:
        return Banana(13)

    # ERROR (expected)
    @staticmethod
    def y(_: T) -> Banana[T]:
        banana = Banana(13)
        return banana


# ERROR (expected)
def z(_: T) -> Banana[T]:
    return Banana(13)
```
Clearly, the returned value has type `Banana[int]` and cannot be generalized to any `Banana[T]`.  And yet, incorrectly, `return Banana(13)` in the first example (`x`) does not raise a type error.

**Expected behavior**
`x` should fail to type-check, same as `y` and `z`.

**VS Code extension or command-line**
command-line, specifically as neovim LSP server.  version 1.1.188.

**Additional context**

If you use a _different_ `TypeVar` than the one defining the class, then an error is correctly raised:
```python
import typing
from __future__ import annotations

T = typing.TypeVar("T")
U = typing.TypeVar("U")


class Banana(typing.Generic[T]):
    stuff: T

    def __init__(self, thing: T):
        self.stuff = thing

    # ERROR (expected behavior)
    @staticmethod
    def x(_: U) -> Banana[U]:
        return Banana(13)
```

However, the behavior should not be different, since staticmethods should not depend on the parameter type of the containing class (which is merely a namespace).  Or, even if it does depend on the containing class's parameter type, the behavior between the two examples `x` and `y` should be consistent.

**Comment by erictraut** [comment]
I agree this is a bug (a false negative). It's likely due to the bidirectional type inference logic used for return statements. I'll investigate further.

The statement "staticmethods should not depend on the parameter type of the containing class" isn't quite correct. The TypeVar of the containing class is scoped by the class, and it's possible to invoke a static method from a specialized version of the class. You can see this in the following:

```python
Banana[int].x(0) # Valid
Banana[int].x("") # Error because str isn't compatible with int
```

**Comment by erictraut** [comment]
This will be fixed in the next release.

**Comment by erictraut** [comment]
This is included in pyright 1.1.198, which I just published. It will also be included in the next release of pylance.
